### PR TITLE
[`flake8-pyi`] Implement PYI018

### PR DESCRIPTION
**Pull request by LaBatata101**
## Summary

Check for unused private `TypeVar`.  See [original implementation](https://github.com/PyCQA/flake8-pyi/blob/2a86db8271dc500247a8a69419536240334731cf/pyi.py#L1958).

```
$ flake8 --select Y018 crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi

crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi:4:1: Y018 TypeVar "_T" is not used
crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi:5:1: Y018 TypeVar "_P" is not used
```

```
$ ./target/debug/ruff --select PYI018 crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi --no-cache

crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi:4:1: PYI018 TypeVar `_T` is never used
crates/ruff/resources/test/fixtures/flake8_pyi/PYI018.pyi:5:1: PYI018 TypeVar `_P` is never used
Found 2 errors.
```
In the file `unused_private_type_declaration.rs`, I'm planning to add other rules that are similar to `PYI018` like the `PYI046`, `PYI047` and `PYI049`.

I'm not happy with all this nesting that is happening here:
https://github.com/astral-sh/ruff/blob/a8377770b21706ad82bee7f1230ef5e9d2ddaa12/crates/ruff/src/checkers/ast/mod.rs#L4486-L4502
I couldn't figure it out, a way to eliminate it.

ref #848

## Test Plan

Snapshots and manual runs of flake8.

### CODE REVIEW

File: crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI018_PYI018.pyi.snap [hunk 1]
```diff
@@ -0,0 +1,22 @@
+---
+source: crates/ruff/src/rules/flake8_pyi/mod.rs
+---
+PYI018.pyi:4:1: PYI018 TypeVar `_T` is never used
+  |
+2 | from typing import TypeVar
+3 | 
+4 | _T = typing.TypeVar("_T")
+  | ^^ PYI018
```
Comment (zanieb) [inline crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI018_PYI018.pyi.snap:9]: Should we highlight the entire declaration?

File: crates/ruff/src/rules/flake8_pyi/rules/unused_private_type_definition.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#[violation]
+pub struct UnusedPrivateTypeVar {
+    name: String,
+}
+
+impl Violation for UnusedPrivateTypeVar {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let UnusedPrivateTypeVar { name } = self;
+        format!("TypeVar `{name}` is never used")
```
Comment (zanieb) [inline crates/ruff/src/rules/flake8_pyi/rules/unused_private_type_definition.rs:37]: Should we note that the variable is private?

File: crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI018_PYI018.pyi.snap (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI018_PYI018.pyi.snap:9]: I personally think highlighting just the symbol is ok.

**Review by charliermarsh — CHANGES_REQUESTED** [review]
Thank you! I know this follows the suggestions I gave in Discord, but looking at the changes and those involved in the downstream PRs, I want to make a few recommendations for modifying the overall data flow. (If you don't feel up for making these changes given the flip-flopping feedback I've given you, I'd also understand and can make them myself if you prefer.)

Overall, it's probably better to avoid (1) having to create a `BindingFlag` for every unused thing we might care about, and (2) having to add all the detection code ("Is this an assignment to a private type var?") to the semantic model itself.

Instead, I'd like to structure it as follows:

- We can create a single `BindingFlag`, something like `PRIVATE_VARIABLE`, and set that within `SemanticModel#add_binding` (rather than having to set flags at all the `add_binding` call sites throughout `checkers/ast/mod.rs`). This flag would just check if the name is underscore-prefixed.
- In `unused_private_type_var`, check if the kind is an unused assignment to a private variable. If so, check if it's an assignment to a private `TypeVar` using the code you already wrote in `handle_node_store` (you can access the statement that defines the binding via `binding.parent`).

This should also generalize to the other variants we care about, without requiring any changes to the semantic model, only the rules themselves. What do you think?

**Comment by LaBatata101** [comment]
> Thank you! I know this follows the suggestions I gave in Discord, but looking at the changes and those involved in the downstream PRs, I want to make a few recommendations for modifying the overall data flow. (If you don't feel up for making these changes given the flip-flopping feedback I've given you, I'd also understand and can make them myself if you prefer.)
> 
> Overall, it's probably better to avoid (1) having to create a `BindingFlag` for every unused thing we might care about, and (2) having to add all the detection code ("Is this an assignment to a private type var?") to the semantic model itself.
> 
> Instead, I'd like to structure it as follows:
> 
>     * We can create a single `BindingFlag`, something like `PRIVATE_VARIABLE`, and set that within `SemanticModel#add_binding` (rather than having to set flags at all the `add_binding` call sites throughout `checkers/ast/mod.rs`). This flag would just check if the name is underscore-prefixed.
> 
>     * In `unused_private_type_var`, check if the kind is an unused assignment to a private variable. If so, check if it's an assignment to a private `TypeVar` using the code you already wrote in `handle_node_store` (you can access the statement that defines the binding via `binding.parent`).
> 
> 
> This should also generalize to the other variants we care about, without requiring any changes to the semantic model, only the rules themselves. What do you think?

That sounds good, I'll implement it myself. I actually thought about doing something more generic like the `PRIVATE_VARIABLE`  flag, but since I'm not that familiar with the code base yet, and I didn't know how to go about it.

**Comment by charliermarsh** [comment]
Great! I renamed to `PRIVATE_DECLARATION` which feels more appropriate when applied to classes and such. I also moved the flag setting into `add_binding`.

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**